maitake/src/wait/cell.rs

```diff
@@ -56,26 +91,24 @@ impl WaitCell {
 }
 
 impl WaitCell {
-    pub fn poll_wait(&self, waker: &Waker) -> Poll<WaitResult<()>> {
+    fn register_wait(&self, waker: &Waker) -> Result<(), Error> {
         tracing::trace!(wait_cell = ?fmt::ptr(self), ?waker, "registering waker");
 
         // this is based on tokio's AtomicWaker synchronization strategy
         match test_dbg!(self.compare_exchange(State::WAITING, State::PARKING, Acquire)) {
             // someone else is notifying, so don't wait!
             Err(actual) if test_dbg!(actual.is(State::CLOSED)) => {
-                return wait::closed();
+                return closed();
             }
             Err(actual) if test_dbg!(actual.is(State::NOTIFYING)) => {
-                waker.wake_by_ref();
-                crate::loom::hint::spin_loop();
-                return wait::notified(());
+                return busy();
```

Comment:
Here we declare that the waitcell is busy, and it is now the callers responsibility to do with that as they will.

Though now that I've made this function private, it means that `Wait` will always pend it's own waker and try again. That might be the wrong behavior, or something we configure in the call to `wake`.

Reply:
i think the problem here is that we have two different "busy" states, the state where another thread is parking a waiter, and the state where the cell is being woken. i don't actually think it's correct to return `busy` in the case where the cell is being woken. in the previous impl, this would wake the waiter, and return `Ready`...i'm not sure if it's correct for the `wait` future to busy-loop forever if it tries to register while the cell is in the middle of waking, i think that should just complete the future, instead?